sdcm/utils/k8s.py

```diff
@@ -609,6 +614,251 @@ def stop(self, timeout=None):
+        - called after a specific Scylla pod IP change
+        - called after any Scylla pod IP change in a specific namespace
+        - called after any Scylla pod IP change in a specific namespace
+          getting 'pod_name=%pod_name%' kwarg.
+          Useful for the objects which will be created later in the future.
+
+        Examples of usages:
+
+        # Per-pod
+        - register_callbacks(func1,
+                             namespace=namespace, pod_name='pod-name-1')
+        - register_callbacks([[func2, args, kwargs]],
+                             namespace=namespace, pod_name='pod-name-2')
+        - register_callbacks([func3, [func3, args2, kwargs2]],
+                             namespace=namespace, pod_name='pod-name-3')
+
+        # Per-pod but with automatic addition of a pod name as a kwarg where the key is 'pod_name'
+        - register_callbacks(func,
+                             namespace=namespace,
+                             add_pod_name_as_kwarg=True)
+
+        # Per-namespace
+        - register_callbacks([[func, args, kwargs], [func2, args, kwargs]],
+                             namespace=namespace)
+
+        Per-pod callbacks get run first, then per-namespace ones.
+        """
+        if not callbacks:
+            LOGGER.warning(
+                "No callbacks are provided. Nothing to register. "
+                "namespace='%s', pod_name='%s'",
+                namespace, pod_name)
+            return
+        elif callable(callbacks):
+            callbacks = [callbacks]
+
+        if namespace not in self.mapper_dict:
+            self.mapper_dict[namespace] = {}
+        if not pod_name:
+            pod_name = '__each__'
+        if pod_name not in self.mapper_dict[namespace]:
+            self.mapper_dict[namespace][pod_name] = {'callbacks': []}
+
+        for callback in callbacks:
+            if callable(callback):
+                callback = [callback, [], {}]
+            if (isinstance(callback, (tuple, list))
+                    and len(callback) == 3
+                    and callable(callback[0])
+                    and isinstance(callback[1], (tuple, list))
+                    and isinstance(callback[2], dict)):
+                self.mapper_dict[namespace][pod_name]['callbacks'].append(
+                    (callback[0], callback[1], callback[2], add_pod_name_as_kwarg))
+            else:
+                LOGGER.warning(
+                    "Unexpected type (%s) of the callback: %s. Skipping", type(callback), callback)
+
+    # NOTE: we want it to run and retry always.
+    #       If it crashes then whole test run will become a mess.
+    @retrying(n=987654, sleep_time=10, allowed_exceptions=(Exception, ))
+    def run(self) -> None:
+        try:
+            for line in self._read_stream():
+                self._process_line(line)
+        finally:
+            if hasattr(self.watcher, 'close'):
+                self.watcher.close()
+
+    def stop(self, timeout=None) -> None:
```

Comment:
this doesn't do what's written on the cover, i.e. it joins, and doesn't stop anything.

Reply:
Read above. Not needed, but I will keep it close to what you expect. The thread will always hang on `self.watcher.close()`.